src/metric/FARCompoundMetric.py

```diff
@@ -0,0 +1,373 @@
+from __future__ import annotations
+from src.metric.CompoundMetric import CompoundMetric
+from src.metric.SampleMetricManager import SampleMetricManager
+from src.core.Setupable import SetupMode
+from src.metric.CompoundMetricManager import CompoundMetricManager
+from typing import Any, TYPE_CHECKING
+from src.population.Population import Population
+from src.metric.DatasetSimilaritySampleMetric import DatasetSimilaritySampleMetric
+from src.metric.PopulationSimilaritySampleMetric import PopulationSimilaritySampleMetric
+import numpy as np
+from src.util.AuxUtil import get_file_jar
+import src.metric.MatchingScore as MS
+from tqdm import tqdm
+import matplotlib.pyplot as plt
+
+if TYPE_CHECKING:
+    from src.compound_model.CompoundModelFactory import CompoundModelFactoryContext
+
+FAR_NAME = "FAR"
+DS_VS_DS_FILE_NAME = (
+    lambda ds_name: "_".join([ds_name, "vs", ds_name, "similarity"]) + ".npy"
+)
+
+# SETUP NAMES
+DS_VS_DS_SETUP_NAME = "DATASET_VS_DATASET"
+
+# Threshold constants
+T_MIN = 0
+T_MAX = 10
+T_STEPS = 50
+
+
+class FARCompoundMetric(CompoundMetric):
+    def __init__(
+        self,
+        cmm: CompoundMetricManager,
+        smm: SampleMetricManager,
+    ):
+        """
+        Constructor for FARCompoundMetric class, subclass of the CompoundMetric class.
+
+        Args:
+            cmm (CompoundMetricManager): Manager used by metric. Population and dataset is derived
+                from this manager.
+            smm (SampleMetricManager, optional): Manager used by metric.
+        """
+        super(FARCompoundMetric, self).__init__(FAR_NAME, cmm, smm)
+
+        # Define threshold linspace
+        self._thresholds = self._get_threshold_space()
+
+        # Init storage structure for this metric
+        self._file_jar = get_file_jar()
+        self._far_ds_vs_ds = None
+        self._far = None
+        self._filter_name_dict = dict()
+
+    def reg_setup_modes(self) -> dict[str, SetupMode]:
+        ds_setup_modes = None
+        pop_setup_modes = None
+        for sm in self._smm.get_metric_instances():
+            if isinstance(sm, DatasetSimilaritySampleMetric):
+                ds_setup_modes = sm.reg_setup_modes()
+            elif isinstance(sm, PopulationSimilaritySampleMetric):
+                pop_setup_modes = sm.reg_setup_modes()
+        assert ds_setup_modes is not None
+        assert pop_setup_modes is not None
```

Comment:
asserts can be nice, but make sure that it is easy to understand what they mean, should they occur

a lot of times I believe it is better to raise a proper exception - especially when there is no mention of the assertion in the docstring

Reply:
I added them to the docstring, will think about my use of assert in the future. They are very convenient! :D